engine/evarutil.ml

```diff
@@ -684,15 +685,37 @@ let undefined_evars_of_named_context evd nc =
     nc
     ~init:Evar.Set.empty
 
-let undefined_evars_of_evar_info evd evi =
+(* compute undefined evars, cache results *) 
+let cache_undefined_evars_of_evar_info ev evd evi =
+  let deps =
   Evar.Set.union (undefined_evars_of_term evd (EConstr.of_constr evi.evar_concl))
     (Evar.Set.union
        (match evi.evar_body with
 	 | Evar_empty -> Evar.Set.empty
 	 | Evar_defined b -> undefined_evars_of_term evd (EConstr.of_constr b))
        (undefined_evars_of_named_context evd
```

Comment:
This code assumes that the evar we're looking at is undefined, because of the call to `Evd.update_undefined`. Better replace the `Evar_defined` branch by an assert false.